File: app/src/main/java/com/travlog/android/apps/libs/rx/transformers/TakeWhenTransformer.java
package com.travlog.android.apps.libs.rx.transformers;

import android.support.annotation.NonNull;

import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.ObservableTransformer;

public final class TakeWhenTransformer<S, T> implements ObservableTransformer<S, S> {

    @NonNull
    private final Observable<T> when;

    public TakeWhenTransformer(final @NonNull Observable<T> when) {
        this.when = when;
    }

    @Override
    @NonNull
    public ObservableSource<S> apply(Observable<S> upstream) {
        return when.withLatestFrom(upstream, (__, x) -> x);
    }
}

File: app/src/main/java/com/travlog/android/apps/libs/rx/transformers/TakeWhenTransformer.kt
package com.travlog.android.apps.libs.rx.transformers

import io.reactivex.Observable
import io.reactivex.ObservableSource
import io.reactivex.ObservableTransformer
import io.reactivex.functions.BiFunction

class TakeWhenTransformer<S, T>(private val `when`: Observable<T>) : ObservableTransformer<S, S> {

    override fun apply(upstream: Observable<S>): ObservableSource<S> {
        return `when`.withLatestFrom(upstream, BiFunction { _, x -> x })
    }
}
